### PR TITLE
Rotating furnitures

### PR DESCRIPTION
**Pull request by Ohda**
**EDITED** This PR allow the rotation of the furnitures.
If the furniture got  `<CanRotate>true</CanRotate>` in the XML, the player can press R and T to rotate the furniture when placing it.

Only Beds and solar panel have been allowed to rotate for now as most of our sprites don't support it.
(shadow issue, perspective issue.

The feature works with save/loading system.
The next step is to manage rotation of Jobspot, workspot.

### CODE REVIEW

**Comment by Tranberry** [comment]
I'm not sure any square (1x1u, 2x2u..) needs rotation. This will force us to add 1-3 more sprites to each furniture or revisit the discussion in https://github.com/TeamPorcupine/ProjectPorcupine/issues/1081 .

The artist are few and far between here, and adding more work to this will hinder some progress as things that will break with art won't become apparent in a speedy manner.

**I do think this is a vital feature,** but maybe not allow all furnitures to be rotated until we got some artists actually replacing the needed art (like the bed).

**Comment by alexfriesen** [comment]
the rotation of the bed looks wrong

**Comment by Raketfart** [comment]
We definetly want to be able to rotate some of the furniture - defined i xml. I imagine some furniture has two rotations, and some have four. I agree that most square furniture doesn't need to be rotated.

Most often, we probably don't want the sprite to just rotate, but instead set a "facing/orientation" on the furniture, and then GetSpriteName should fetch, and maybe flip the right sprite. When rotating we should switch the width and height of the furniture, so 2x3 becomes 3x2. 

Solar panels look ok when rotated, if the size of the furniture gets fixed.

**Comment by sboigelot** [comment]
@Raketfart you would rotate square furniture to face the workspots inputs and outputs

**Comment by Tranberry** [comment]
@sboigelot I believe @Raketfart meant that there would be independent sprites on the sprite-sheet for each facing.

**Comment by Raketfart** [comment]
@sboigelot I Agree. I meant things like the RTG or the landing pad, that take up the same space, and has no functional or cosmetic difference if rotated.

**Comment by Ohda** [comment]
If i can finsih this i think i will only add rotation on bed and solar panel for now.

**Comment by Ohda** [comment]
Placement problem are fixed.

Still looking forward for the saving/loading issues.

**Comment by Ohda** [comment]
Remove the WIP tag.

- The furniture rotation is bind to R and T Key.
- If the furniture got <CanRotate>true</CanRotate> in the XML, the rotation is allowed when placing the furniture.
- Bed and Solar panel are active for rotation.
- The furnitures rotation are saved/Loaded from the save file.

File: Assets/Scripts/Models/Area/World.cs [hunk 1]
```diff
@@ -194,6 +194,7 @@ public Furniture PlaceFurniture(string type, Tile t, bool doRoomFloodFill = true
         }
 
         Furniture furn = PrototypeManager.Furniture.Get(type);
+        furn.SetRotation(rotation);
```
Comment (TomMalbran) [inline Assets/Scripts/Models/Area/World.cs]: I think that we should clone the furniture before setting the rotation.

File: Assets/Scripts/Models/Area/World.cs [hunk 2]
```diff
@@ -240,10 +241,11 @@ public Furniture PlaceFurniture(Furniture furniture, Tile t, bool doRoomFloodFil
...
     public void JobComplete_FurnitureBuilding(Job theJob)
     {
-        // Let our workspot tile know it is no longer reserved for us
-        UnreserveTileAsWorkSpot((Furniture)theJob.buildablePrototype, theJob.tile);
+        Furniture furn = (Furniture)theJob.buildablePrototype;
 
-        PlaceFurniture(theJob.JobObjectType, theJob.tile);
+        // Let our workspot tile know it is no longer reserved for us
+        UnreserveTileAsWorkSpot(furn, theJob.tile);
+        PlaceFurniture(theJob.JobObjectType, theJob.tile, true, furn.Rotation);
```
Comment (TomMalbran) [inline Assets/Scripts/Models/Area/World.cs]: should we just use furn at this point? is already a clone... and we might not even need to clone it again. What do you think?

Comment (Ohda) [inline Assets/Scripts/Models/Area/World.cs]: Not sure why i was afraid to doing it. Work like a charm.

File: Assets/Scripts/Controllers/Sprites/JobSpriteController.cs [hunk 3]
```diff
@@ -122,7 +122,8 @@ protected override void OnCreated(Job job)
                 sr.sprite = usc.GetSpriteForUtility(job.JobObjectType);
             }
 
-            job_go.transform.position = job.tile.Vector3 + ImageUtils.SpritePivotOffset(sr.sprite);
+            job_go.transform.position = job.tile.Vector3 + ImageUtils.SpritePivotOffset(sr.sprite, BuildModeController.Instance.GetCurrentPreviewRotation());
```
Comment (TomMalbran) [inline Assets/Scripts/Controllers/Sprites/JobSpriteController.cs]: Should we only rotate furniture, and maybe use `job.buildablePrototype.Rotation` (if we only check for furniture)?

File: Assets/Scripts/Models/Area/World.cs (hunk 2 above)
Comment (TomMalbran) [inline Assets/Scripts/Models/Area/World.cs]: Now that furn is no longer the prototype, it does work.

**Comment by Ohda** [comment]
All suggestions from @TomMalbran as been implented. Thanks you !

**Comment by koosemose** [comment]
Conflicted

**Comment by Ohda** [comment]
Conflict resolved

**Comment by dusho** [comment]
so in the newest version only bed and solar panels are rotate-able? From what I tried, all works ok.
Would be curious how smelter and other stuff would work - probable can't since offsets are placed specifically in LUA?

**Comment by TomMalbran** [comment]
They should all work. But the graphics look bad when rotating some furniture. We would need to create multiple sprites for the rotations that we want.

The offsets are not set in Lua, but the offset vectors will need to be changed so that it works for the rotations. Is totally doable though.

**Comment by Ohda** [comment]
I'll try to add a Jobspot to the bed to work on the rotation of thoses spots that need rotation.
Can it be another PR ? I'am afraid of the big one !

**Comment by koosemose** [comment]
This does not seem to be working anymore, unless you changed the rotation keys from B and N, in which case please change it in the top post for easier testing.

**Comment by Ohda** [comment]
Yes this is R and T now as suggested.
Top post edited

**Comment by koosemose** [comment]
Looks good to me, Stylecop clean. Merged.

**Pull request merged**